x/ccv/provider/keeper/keeper.go

```diff
@@ -1184,3 +1184,141 @@ func (k Keeper) IsOptIn(ctx sdk.Context, chainID string) bool {
 	topN, found := k.GetTopN(ctx, chainID)
 	return !found || topN == 0
 }
+func (k Keeper) SetOptedIn(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+	blockHeight uint64,
+) {
+	store := ctx.KVStore(k.storeKey)
+
+	// validator is considered opted in
+	blockHeightBytes := make([]byte, 8)
+	binary.BigEndian.PutUint64(blockHeightBytes, blockHeight)
+
+	store.Set(types.OptedInKey(chainID, providerAddr), blockHeightBytes)
+}
+
+func (k Keeper) DeleteOptedIn(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.OptedInKey(chainID, providerAddr))
+}
+
+func (k Keeper) IsOptedIn(
+	ctx sdk.Context,
+	chainID string,
+	providerAddr types.ProviderConsAddress,
+) bool {
+	store := ctx.KVStore(k.storeKey)
+	return store.Get(types.OptedInKey(chainID, providerAddr)) != nil
```

Comment:
The block height would always be smaller or equal to the current block height. It would be equal to the current block height when we first opt in that validator, for instance at the end of an epoch.